Add CloudInit controller

connorkuehl opened the pull request:
~~Depends on: #14~~

---

**Problem:**
<!-- Explain the problem you are aiming to resolve in this PR. -->

A Harvester cluster operator can currently modify node configuration by SSH'ing into the node, and modifying the relevant Elemental cloud-init configuration file under `/oem`. However, this is incompatible with a GitOps approach to managing the state of the cluster, at least, without writing some code of their own.

**Solution:**
<!-- Example: When "Adding a function to do X", explain why it is necessary to have a way to do X. -->
This PR adds a new CloudInit CRD, a controller, and an fsnotify watcher to harvester-node-manager. This allows cluster operators to layer a cloud-init configuration on top and to target specific nodes in the cluster with the `matchSelector` field in the cloud-init spec.

For example, to add SSH access to the Rancher user with my SSH key across every node in the cluster, I can represent this like so:

```
apiVersion: node.harvesterhci.io/v1beta1
kind: CloudInit
metadata:
  name: ssh-access
spec:
  matchSelector: {}
  filename: 99_ssh.yaml
  contents: |
    stages:
      network:
        - authorized_keys:
            rancher:
              - ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIPCUNsQEnKj0nl1GS07Qr5RDCCbCim4wu06hCzQZDmTk ckuehl@suselaptop
```

**Related Issue:** https://github.com/harvester/harvester/issues/3902

**Test plan:**
<!-- Make sure tests pass on the Circle CI. -->
Much of the test plan appears in this PR as automated unit tests.

However, to confirm operation on a multi-node cluster, I ran these manual steps:

1. Apply the CRD (`kubectl apply -f ./manifests/crds/node.harvesterhci.io_cloudinits.yaml`)
2. Import a container built from this PR onto all nodes in the cluster (`ctr -n=k8s.io image import /home/rancher/harvester-node-manager.tar`)
3. Patch the Harvester ManagedChart (`kubectl patch managedchart harvester -n fleet-local -p='{"spec":{"paused":true}}' --type=merge`)
4. Edit the harvester-node-manager deployment to use the container imported in step 2 (`kubectl edit daemonset/harvester-node-manager -n harvester-system`)
5. Rollout the new container (`kubectl rollout restart daemonset/harvester-node-manager -n harvester-system`)
6. Wait for that to finish
7. Apply a CloudInit object (like the one listed above)
8. SSH in, confirm the file has been created under `/oem`.

connorkuehl commented:
I'm wondering if it's worth writing a follow-up PR to add an admission controller that ensures the document is parsable with `yip` as well as protecting certain files, like `elemental.config`, `grubenv`, `harvester.config`, `install`.

connorkuehl commented:
ping

Vicente-Cheng commented:
Hi @connorkuehl, I will take a look at this PR next week. Sorry for the late update.

connorkuehl commented:
ping

Vicente-Cheng commented on pkg/monitor/cloudinit.go:
```diff
@@ -0,0 +1,97 @@
+
+	path := event.Name
+
+	cloudinits, err := m.cloudinitsCache.List(labels.Everything())
```
if `label` is nil, we will get `label.Everything()`
Maybe just `m.cloudinitsCache.List()`

connorkuehl replied:
🤔  I'm not sure I'm following this comment. cloudinitsCache.List requires an argument. Are you suggesting passing `nil` there?

Vicente-Cheng replied:
Ah, sorry. I have some misunderstanding here. Using `labels.Everything()` is correct.

Vicente-Cheng commented on pkg/monitor/cloudinit.go:
```diff
@@ -0,0 +1,97 @@
+	}()
+}
+
+func (m *CloudInitMonitor) handleFSNotify(event fsnotify.Event) {
```
Just have some questions about this event-driven case.
I saw we will return with some error path. Could we have a chance to retry?

Vicente-Cheng commented on pkg/cloudinit/cloudinit.go:
```diff
@@ -0,0 +1,63 @@
+
+var Directory = "/host/oem"
+
+func RequireLocal(cloudinit *harvesterv1beta1.CloudInit) error {
```
I thought this function applied cloudinit to local. Could we add some comments for it or make the function name more clear?

connorkuehl replied:
> I thought this function applied cloudinit to local.

It does 🙂 

> Could we add some comments for it or make the function name more clear?

Yeah I'll add a comment for the next revision

connorkuehl commented on pkg/cloudinit/cloudinit.go:
```diff
@@ -0,0 +1,63 @@
+
+const AnnotationHash = "harvesterhci.io/hash"
+
+var Directory = "/host/oem"
```
This package is consumed by two other packages, so I'd have to leave it exported.

And when it's used elsewhere, it reads `cloudinit.Directory`, which comes across a little more clear to me than `cloudinit.OEMDir`.

If this doesn't sway you, let me know and I'll change it 🙂

Vicente-Cheng replied:
OK, makes sense to me. Thanks for the explanation.

connorkuehl commented on pkg/controller/cloudinit/controller.go:
```diff
@@ -0,0 +1,69 @@
+		nodeCache:  nodeCache,
+	}
+
+	cloudinits.OnChange(ctx, handlerName, ctl.OnCloudInitChange)
```
When I was first ruminating on this with @ibrokethecloud we initially decided against this... but I could go either way. Gaurav, do you have a stronger opinion on this?

ibrokethecloud replied:
we could add the `OnRemove` case but I am not sure there is an easy way to rollback the impact of the changes from the underlying host.

may be at best `OnRemove` cleans up the file from the underlying host, but any operational changes would need to be done by another cloud-init object which reverses the effects?

connorkuehl replied:
We could have OnRemove remove the file, then once the node is rebooted it won't have been applied.

connorkuehl commented:
Thanks for the review, @Vicente-Cheng! 🙂 For the comments I have left open and haven't responded to, I am working on addressing in the next revision.

ibrokethecloud commented on pkg/monitor/cloudinit.go:
```diff
@@ -0,0 +1,97 @@
+		return
+	}
+
+	logrus.Infof("Managed file %s has changed on disk, overwriting", event.Name)
```
may be we should just requeue the CRD if the file has changed, and let the CRD fall through the controller reconcile?

This way the file is only ever changed by the controller reconcile logic.

connorkuehl commented on pkg/apis/node.harvesterhci.io/v1beta1/cloudinit.go:
```diff
@@ -0,0 +1,22 @@
+type CloudInit struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              CloudInitSpec `json:"spec"`
```
@Vicente-Cheng This is my first "from-scratch" controller, so apologies if this is a naive question.

Since the CloudInit object might not apply to all nodes, would it be OK if the status conditions were "dynamic" based on the node's name?

For example, a condition name would be formatted like `fmt.Sprintf("AppliedTo%s", nodeName)` with a status of "True", "False", or "Unknown" ?

Or is it expected that condition names are static and well-known as part of the controller's behavior?

connorkuehl replied:
@Vicente-Cheng ping

Vicente-Cheng replied:
Sorry @connorkuehl, for late reply.
I think you could try to check the `Status.Condition` related article on the k8s.
That would help people to know the status of this CRD.

connorkuehl replied:
Latest revision consists of a Status member which describes the CloudInit object relevant to each node in the cluster:

```
Status:
  Rollouts:
    Conditions:
      Last Transition Time:  2023-10-24T21:15:02Z
      Message:               
      Reason:                CloudInitApplicable
      Status:                True
      Type:                  Applicable
      Last Transition Time:  2023-10-24T21:09:28Z
      Message:               Local file checksum is the same as the CloudInit checksum
      Reason:                CloudInitChecksumMatch
      Status:                False
      Type:                  OutOfSync
      Last Transition Time:  2023-10-24T21:09:28Z
      Message:               99_ssh.yaml is present under /oem
      Reason:                CloudInitPresentOnDisk
      Status:                True
      Type:                  Present
    Node Name:               harv1
```

connorkuehl commented:
ping

Vicente-Cheng commented on pkg/controller/cloudinit/controller.go:
```diff
@@ -0,0 +1,231 @@
+	}
+
+	if cloudinit.MatchesNode(node, cloudInitCopy) {
+		_ = cloudinit.RequireLocal(cloudInitCopy)
```
Should we need to do the error handling here?
Or do we have another mechanism to ensure the cloudinit is synced with local?

connorkuehl replied:
Yeah, I think this should call update status and then return the error so it gets requeued

Vicente-Cheng commented on pkg/monitor/cloudinit.go:
```diff
@@ -0,0 +1,107 @@
+		return
+	}
+
+	logrus.Debugf("Fetched node")
```
edundant log? We could check the following `"Fetched %d CloudInits for %s"`

Vicente-Cheng reviewed (commented):
overall lgtm, but we still need to discuss the `OnRemove` part.

connorkuehl commented on go.mod:
```diff
@@ -45,6 +45,13 @@ require (
 )
 
+require (
+	github.com/fsnotify/fsnotify v1.6.0
```
It's part of go-common's public interface: https://github.com/harvester/go-common/blob/e2c0c177ddd765496a123f70c5b0e188734c099f/watcher.go#L51

Vicente-Cheng replied:
Got it! Thanks for the explaination

connorkuehl commented on pkg/controller/cloudinit/controller.go:
```diff
@@ -0,0 +1,231 @@
+	}
+
+	checksum, err := cloudinit.Measure(strings.NewReader(cloudInitCopy.Spec.Contents))
+	_ = err
```
Yeah, will fix

ibrokethecloud commented:
@Vicente-Cheng @connorkuehl i dont think we should be second guessing the rollback by performing some operations during `OnRemove`

afaik, apart from the chroot stage, all commands are applied on top of the base image but not persisted to the image itself. They just get re-applied each time on boot. So removing the file should rollback the effect of those changes?

We should advise the users that to rollback the effect of earlier changes they will need to manually submit another CRD which contains steps to reverse the original behavior.

connorkuehl commented:
> We should advise the users that to rollback the effect of earlier changes they will need to manually submit another CRD which contains steps to reverse the original behavior.

@ibrokethecloud  @Vicente-Cheng  The controller doesn't automatically apply the changes, right now it assumes the user will reboot the node after making changes to the CRD, so under that model just removing the CRD and rebooting the node should be sufficient, right?

w13915984028 commented on pkg/cloudinit/cloudinit.go:
```diff
@@ -0,0 +1,68 @@
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tempFile.Name())
```
should those 2 still be deferred after `os.Rename` ?

 `tempFile.Close()`  and then `os.Rename` ?

connorkuehl replied:
It is harmless during the success path and it helps tidy things up if the io.Copy fails.

w13915984028 commented on pkg/cloudinit/cloudinit.go:
```diff
@@ -0,0 +1,68 @@
+// the given `cloudinit` object is an exact copy of the `cloudinit` object's
+// contents.
+func RequireLocal(cloudinit *cloudinitv1.CloudInit) error {
+	absPath := filepath.Join(Directory, cloudinit.Spec.Filename)
```
Use webhook or `while-list` to control the allowed file will be more secure now.

connorkuehl replied:
Webhook has since merged to the repo which helps secure this

w13915984028 reviewed (commented):
Thanks for adding this new feature.

A few questions / suggestions:

(1) Add an EPIC to include:
  New CRD & controller deserve a HEP to describe the user-story, specification, test plan ...
  Update Harveter DOC about feature usage, troubleshooting ...
  Potential UI enhancement
  Possible webhook to check file name / format

(2) Those internal configuration files are very important for the Harvester cluster to work correctly, malformed files may cause cluster to fail to start. This feature will expose those files to be manipulated officially, it is important to make sure the exposed file path & name & contents, at the first stage, allowing specific files will be more secure.

Vicente-Cheng commented:
>@ibrokethecloud @Vicente-Cheng The controller doesn't automatically apply the changes, right now it assumes the user will reboot the node after making changes to the CRD, so under that model just removing the CRD and rebooting the node should be sufficient, right?

Hi @connorkuehl,
Yes, I thought that would be enough. Does this improvement would have the UI part? If yes, maybe we need to add some tips when the user deletes the cloud-init config.

ibrokethecloud commented on pkg/controller/cloudinit/controller.go:
```diff
@@ -0,0 +1,231 @@
+		return cloudInitCopy, err
+	}
+
+	if cloudinit.MatchesNode(node, cloudInitCopy) {
+		_ = cloudinit.RequireLocal(cloudInitCopy)
+	}
+
+	return c.updateStatus(node, cloudInitCopy)
```
we should re-organise this a bit. If a node does not match the condition, we should return earlier.

currently when a node is not matching the filters, even then we are going through the process of evaluating all the conditions.

if notApplicable condition is set in the first method,  https://github.com/harvester/node-manager/pull/15/files#diff-8ad79d727a2364fb392ad1a30a059ed49322057d1bcc8788eefb0f83737ba154R107-R111, it would be safe to assume none of the other conditinos will ever be true. may be wise to apply those conditions and update status and return quickly?

connorkuehl replied:
Agreed, will incorporate that.

connorkuehl replied:
It occurred to me when I went to update this during the rebase that there is a situation in which we'd want to process all of the conditions, and ultimately I think it is simplest to do so unconditionally.

For example if the CloudInit resource's MatchSelector is updated and that results in one or more nodes being excluded, all 3 conditions can be used as valuable hints to the operator to know why a node is still being customized a certain way.

I.e., Applicable == False, but Present == True suggests the controller's automatic removal failed and the operator can get things back the way they expect it to by SSH'ing in and manually removing it, etc.

ibrokethecloud commented on pkg/apis/node.harvesterhci.io/v1beta1/cloudinit.go:
```diff
@@ -0,0 +1,41 @@
+}
+
+type CloudInitStatus struct {
+	Rollouts []Rollout `json:"rollouts,omitempty"`
```
looking at the way rollouts are evaluated in the controller would it be better to make `Rollouts` a `map[nodeName]RollOut`? 

since the node-manager is going to run as a daemonset, it makes it easier for each controller instance to just lookup Rollout status using its node name and perform updates to only the element in the map?

connorkuehl replied:
Absolutely, will update.

ibrokethecloud commented:
Other than the minor feedback, the changes work as expected. I was able to deploy a custom cloud-init to a 2 node cluster.

sample cloud-init is as follows:

```
apiVersion: node.harvesterhci.io/v1beta1
kind: CloudInit
metadata:
  name: sample-cloud-init
spec:
  contents: |
    #cloud-config
    write_files:
    - encoding: b64
      content: CiMgVGhpcyBmaWxlIGNvbnRyb2xzIHRoZSBzdGF0ZSBvZiBTRUxpbnV4
      path: /tmp/content
      permissions: "0644"
  matchSelector:
    "kubernetes.io/hostname": "node2"
  filename: "write-sample.yml"
```

```
k get nodes
NAME    STATUS   ROLES                       AGE     VERSION
node1   Ready    control-plane,etcd,master   3d20h   v1.26.11+rke2r1
node2   Ready    <none>                      3d20h   v1.26.11+rke2r1
```

cloud-init reconciles correctly, and can see changes on node2 only

```
node2:/oem # ls -lart
total 33
drwx------   2 root root 12288 Dec 14 03:24 lost+found
-rw-rw-rw-   1 root root  9463 Dec 14 03:24 90_custom.yaml
-rw-------   1 root root  1747 Dec 14 03:25 harvester.config
-rwxr-xr-x   1 root root   585 Dec 14 03:25 elemental.config
drwxr-xr-x. 22 root root  4096 Dec 14 03:35 ..
-rw-r--r--   1 root root  1024 Dec 14 03:35 grubenv
drwxr-xr-x   2 root root  1024 Dec 14 03:35 install
-rw-------   1 root root   154 Dec 17 23:51 write-sample.yml
drwxr-xr-x   4 root root  1024 Dec 17 23:51 .
node2:/oem # more write-sample.yml
#cloud-config
write_files:
- encoding: b64
  content: CiMgVGhpcyBmaWxlIGNvbnRyb2xzIHRoZSBzdGF0ZSBvZiBTRUxpbnV4
  path: /tmp/content
  permissions: "0644"
node2:/oem #
```

no changes are deployed to node1

```
node1:/oem # ls -alrt
total 119
drwx------   2 root root 12288 Dec 14 02:56 lost+found
-rw-rw-rw-   1 root root 98248 Dec 14 02:57 90_custom.yaml
-rw-------   1 root root  1754 Dec 14 02:57 harvester.config
-rwxr-xr-x   1 root root   586 Dec 14 02:57 elemental.config
drwxr-xr-x. 22 root root  4096 Dec 14 03:06 ..
-rw-r--r--   1 root root  1024 Dec 14 03:06 grubenv
drwxr-xr-x   2 root root  1024 Dec 14 03:06 install
-rw-r--r--   1 root root   238 Dec 14 03:15 99_settings.yaml
drwxr-xr-x   4 root root  1024 Dec 14 03:15 .
```

cloudinit object reflects the status correctly:

```
status:
  rollouts:
  - conditions:
    - lastTransitionTime: "2023-12-17T23:51:01Z"
      message: MatchSelector does not match Node labels
      reason: CloudInitNotApplicable
      status: "False"
      type: Applicable
    - lastTransitionTime: "2023-12-17T23:51:01Z"
      message: Local file checksum is different than the CloudInit checksum
      reason: CloudInitChecksumMismatch
      status: "True"
      type: OutOfSync
    - lastTransitionTime: "2023-12-17T23:51:01Z"
      message: write-sample.yml is absent from /oem
      reason: CloudInitAbsentFromDisk
      status: "False"
      type: Present
    nodeName: node1
  - conditions:
    - lastTransitionTime: "2023-12-17T23:51:01Z"
      message: ""
      reason: CloudInitApplicable
      status: "True"
      type: Applicable
    - lastTransitionTime: "2023-12-17T23:51:01Z"
      message: Local file checksum is the same as the CloudInit checksum
      reason: CloudInitChecksumMatch
      status: "False"
      type: OutOfSync
    - lastTransitionTime: "2023-12-17T23:51:01Z"
      message: write-sample.yml is present under /oem
      reason: CloudInitPresentOnDisk
      status: "True"
      type: Present
    nodeName: node2
```

modifications to the file are detected and reconciled as expected

```
node2:/oem # > write-sample.yml
node2:/oem # ls -alrt
total 33
drwx------   2 root root 12288 Dec 14 03:24 lost+found
-rw-rw-rw-   1 root root  9463 Dec 14 03:24 90_custom.yaml
-rw-------   1 root root  1747 Dec 14 03:25 harvester.config
-rwxr-xr-x   1 root root   585 Dec 14 03:25 elemental.config
drwxr-xr-x. 22 root root  4096 Dec 14 03:35 ..
-rw-r--r--   1 root root  1024 Dec 14 03:35 grubenv
drwxr-xr-x   2 root root  1024 Dec 14 03:35 install
-rw-------   1 root root   154 Dec 17 23:55 write-sample.yml
drwxr-xr-x   4 root root  1024 Dec 17 23:55 .
node2:/oem # more write-sample.yml
#cloud-config
write_files:
- encoding: b64
  content: CiMgVGhpcyBmaWxlIGNvbnRyb2xzIHRoZSBzdGF0ZSBvZiBTRUxpbnV4
  path: /tmp/content
  permissions: "0644"
```

when a file is removed and resynced by the controller, should the timestamp on the conditions change to reflect the last action? currently the condition timestamps are not updated.

connorkuehl commented:
@ibrokethecloud thanks for the review! I will incorporate them after I get the webhook patches merged. :slightly_smiling_face:

connorkuehl commented:
Hey everyone,

Thanks for sticking with me :sweat_smile:. The HEP has been reviewed, and the webhook that fell out of it has been merged.

The notable changes to this patch since you last saw it is directly addressing feedback, but there are a couple of behavioral changes that emerged since the last version you've reviewed due to requirements that we drew from the HEP.

The first change is that the controller checks the CloudInit Spec to see if it has Paused: true. If so, it does not attempt to reconcile, it just makes sure the Status is up-to-date.

The second change is that the controller now creates Events when it overwrites or removes a file from a Harvester host. It looks a little something like this:

```
Events:
  Type    Reason                  Age                From                    Message
  ----    ------                  ----               ----                    -------
  Normal  CloudInitFileModified   18m (x2 over 18m)  harvester-node-manager  99_ssh.yaml has been overwritten on harv1
  Normal  CloudInitNotApplicable  2m5s               harvester-node-manager  99_ssh.yaml has been removed from harv1
  Normal  CloudInitNotApplicable  2m5s               harvester-node-manager  99_ssh.yaml has been removed from harv2
  Normal  CloudInitNotApplicable  2m5s               harvester-node-manager  99_ssh.yaml has been removed from harv3
  Normal  CloudInitFileModified   92s                harvester-node-manager  99_ssh.yaml has been overwritten on harv2
  Normal  CloudInitFileModified   92s                harvester-node-manager  99_ssh.yaml has been overwritten on harv3
  Normal  CloudInitFileModified   79s (x2 over 92s)  harvester-node-manager  99_ssh.yaml has been overwritten on harv1
```

Note that if you want to try this out for yourself, it is a little more involved this time around because my PR to enable image builds/pushes of the webhook is not merged yet, and I am actively working on updating node-manager's Helm chart to include the webhook.

That means that if you want to try it out, you'll need to do some setup:

1. Apply the CloudInit CRD https://github.com/harvester/node-manager/raw/master/manifests/crds/node.harvesterhci.io_cloudinits.yaml
1. Apply the updated RBAC in this PR https://github.com/harvester/node-manager/pull/15/files#diff-a3f9e35eaeb026c24daddfa3ae7fce5b17879ad337771f161e655643f1ba1696
1. Apply the webhook deployment: https://github.com/harvester/node-manager/raw/master/manifests/deployment.yaml
1. Apply the webhook service: https://github.com/harvester/node-manager/raw/master/manifests/service.yaml

Then you can check out this branch, run `make build`, `make package`, make `package-webhook` and `docker image save` both of those things, `scp` them to your Harvester node(s) and import them (`ctr -n=k8s.io image import ...`)

and don't forget to update the webhook deployment to use the imported image as well as the node-manager daemonset and then restart their rollouts.

w13915984028 reviewed (approved):
LGTM, thanks.

markhillgit reviewed (approved):
lgtm thanks!

ibrokethecloud commented on pkg/controller/cloudinit/controller.go:
```diff
@@ -0,0 +1,403 @@
+		logrus.WithError(err).
+			WithField("cloudinit_name", cloudInitCopy.Name).
+			Warn("Failed to emit removal event for CloudInit")
+	}
```
is there any specific reason we do not handle the error?

connorkuehl replied:
It seems kind of awkward to add retrying inline since I am not sure if it blocks processing of updates to other objects while this is stuck in a retry loop. 


If I return an err to have the controller retry, the previous iteration has already reconciled the state, so the flow of control won't think it's necessary to emit an event. I could try saving this state in an annotation on the object, but perhaps it's likely that whatever symptom prevented the event write could also prevent the object write? :thinking:

ibrokethecloud replied:
now that i think about it, ignoring may be ok since there is nothing else that can be done about the file other than emit an error and report a condition.

ibrokethecloud commented on pkg/controller/cloudinit/controller.go:
```diff
@@ -0,0 +1,403 @@
+	}
+
+	checksum, err := cloudinit.Measure(strings.NewReader(cloudInitCopy.Spec.Contents))
+	_ = err
```
should we be handling this error?

Vicente-Cheng commented on pkg/controller/cloudinit/controller.go:
```diff
@@ -0,0 +1,403 @@
+		return cloudInitCopy, err
+	}
+
+	if cloudInitCopy.Spec.Paused {
```
I got confused here.

When we paused the cloudinit CR, we skip do anything. Is it what you want?
If yes, could we check it more early? Also, we can return nil.

If we try to update and return the update result, that will trigger the OnChange again.

connorkuehl replied:
> When we paused the cloudinit CR, we skip do anything. Is it what you want?

Right, but more specifically I want the controller to not modify any files on disk,

>  If yes, could we check it more early? Also, we can return nil.

but I still want it to make sure the checksum annotation is correct on the resource and to make sure the status is up-to-date, so we can't return earlier as we need the checksum reconciled as well as the status.

Vicente-Cheng replied:
Monitoring the checksum seems to make sense.
But, IIRC, any change for the cloud-init CR will trigger the OnChange again. So it will check the checksum at that time.

Keep return c.updateStatus here might loop the OnChange.

cc @ibrokethecloud if I got something wrong with that mechanism

ibrokethecloud replied:
the status would not have changed and that will return no status updates and it will reconcile successfully.

Vicente-Cheng replied:
Got it; thanks for the explanation.

Vicente-Cheng commented on pkg/controller/cloudinit/controller.go:
```diff
@@ -0,0 +1,403 @@
+
+		switch cond.Type {
+		case string(cloudinitv1.CloudInitFilePresent):
+			if createdAt.After(cond.LastTransitionTime.Time) {
```
Why do we use `createdAt` instead of the current time?
I think that should not have been much different because you just init createdAt when call into this func

Vicente-Cheng replied:
oops... missing this line
```
_ = unix.Stat(filepath.Join(cloudinit.Directory, cloudInitObj.Spec.Filename), &st)
```

You are right, that make sense to me!

Vicente-Cheng commented on pkg/controller/cloudinit/controller.go:
```diff
@@ -0,0 +1,403 @@
+				cond.LastTransitionTime = metav1.NewTime(createdAt)
+			}
+		case string(cloudinitv1.CloudInitOutOfSync):
+			if modifiedAt.After(cond.LastTransitionTime.Time) {
```
ditto

connorkuehl replied:
I used the file's modified time since that was when it transitioned from OutOfSync==True to OutOfSync==False and vice versa.

Vicente-Cheng replied:
Same as above, thanks!

Vicente-Cheng commented:
Generally works well. Leave some comments on the above.

Tested with your node-manager/node-manager-webhook

test-cloudinit CR
```
apiVersion: node.harvesterhci.io/v1beta1
kind: CloudInit
metadata:
  name: sample-cloud-init
spec:
  contents: |
    #cloud-config
    write_files:
    - content: "Hello cloud-init!"
      path: /tmp/hello.txt
      permissions: "0644"
  matchSelector:
    "kubernetes.io/hostname": "harvester-node-0"
  filename: "sample.yml"
```

Check on the `/oem/` folder
```
harvester-node-0:~ # cat /oem/sample.yml
#cloud-config
write_files:
- content: "Hello cloud-init!"
  path: /tmp/hello.txt
  permissions: "0644"
```

After reboot, check the /tmp/hello.txt
```
harvester-node-0:~ # cat /tmp/hello.txt
Hello cloud-init!
```

Also, check event work as well.
Thanks for working on that!

connorkuehl commented on manifests/rbac.yaml:
```diff
@@ -19,6 +19,9 @@ rules:
     verbs: [ "get", "watch", "list", "update" ]
+  - apiGroups: [ "" ]
+    resources: [ "events" ]
+    verbs: [ "create", "get", "list", "update" ]
```
Right, seems already merged https://github.com/harvester/node-manager/blob/4381ae1126d191bd9a057d0f7f0165b7bd6033fb/manifests/rbac.yaml#L40

Vicente-Cheng replied:
Ah, I know the reason.
The service account should be `harvester-node-manager-webhook` instead `harvester-node-manager`

There are two place you need to modify.
https://github.com/harvester/node-manager/blob/master/manifests/deployment.yaml#L26
https://github.com/harvester/charts/pull/205/files#diff-46be9358543a418d2319c41db8a8098bf0707ceb21fd89706c3f0874e13db7daR19

connorkuehl replied:
Thanks! :slightly_smiling_face: 

https://github.com/harvester/node-manager/pull/30

ibrokethecloud reviewed (approved):
lgtm. thanks.

Vicente-Cheng reviewed (approved):
nice work! thanks!

markhillgit commented:
@connorkuehl you are approved to merge this
